### ISSUE TITLE
Page Title Module in core

### ISSUE THREAD

**Issue by bd0bd** [backdrop/backdrop-issues#1191]
> **quicksketch**:
> The question about modifying the markup of the >h1< heading unfortunately is a bit more complicated to answer. Let's file a new issue as a question to answer it fully if needed.

Old fashioned HTML is far away from modern days
Pages markup structure is much more detailed nowdays.
Two types are actual - h1 and header h1 - for pages.
Also "hidden" option would be good to have (I have this option in the CMS I have used during about 4 years).
Would be good to have a possibility to switch it for example like this (rough draft of the rough idea):

![page-title](https://cloud.githubusercontent.com/assets/13515302/9874538/41a28b30-5baf-11e5-84b9-aafc228669e5.gif)

**Comment by sutibun** [comment]
I don't know about this. The question that immediately comes up are:

- will most people need to change the markup that often that we need to present an UI element for it?
- when encountering this for the first time, will this confuse normal users? Will they even know what it's for?
- why do you need to swap the markup so often?
- isn't this best suited as a contrib module or to be addressed in the site's design?
- if you have to change the markup often, isn't this the fault of the designer?

Don't get me wrong.

I can *definitely* see a use for this. I know because I wanted a similar feature except what I wanted was a custom field which would display custom text below the article tittle and the featured image but above the actual article content.

Why? To get readers in the right frame of mind before reading the article.  I wanted the custom field to sometimes be 

- another sub headline (h2), 
- a short intro/summary introducing the reader to the article (p) or 
- a related quote from a famous person (blockquote)

In other words, manipulate the header.

It would all depend on the article and how I wanted to start it. I wasn't too techie so what I finally went with in the end is creating a custom text field that allowed the use of the markdown format. With markdown enabled there, I could achieve all I wanted. (I could use \#\# for a subhead line or \> for blockquote)

I removed the field years later btw. Looking back now, maybe another approach you could take is creating a different content type, one with h1 another with the other. Or create a custom field like I did.

Anyway, long story short, this features would require the user to know HTML markup. But if you think about it, the reason there's a rich text editor is to hide away the complexity because normal users don't know HTML.

**Comment by MrHaroldA** [comment]
A lot of our websites have to have a correct heading nesting. For example: you can't have an H2 without an H1 higher up in the DOM. Hiding the H1 will invalidate the whole page.

This is one of the many reasons I'm against such a feature.

**Comment by klonos** [comment]
I too cannot think off the top of my head any valid use case where one would allow content authors/editors to decide and alter the site's markup "on th fly" upon content creation. Using different content types seems more appropriate to me ...and if these need to be displayed together (which aesthetically would not be good - especially to people with self-diagnosed OCD like myself), then use a view that includes and lists all these content types.

Would love to hear the use case at hand though before dismissing this. I'm sure there must be a simpler way.

**Comment by bd0bd** [comment]
Backdrop is SEO unfriendly CMS.
We do not have simple options to:

a) make custom >title< (different from >h1<)
b) make custom >h1< (different from >title<)
c) make meta Description
d) make meta Keywords
e) canonical tag

These things are important for sites.
I would like to make normal sites like Nielsen's one: http://www.nngroup.com/

**Comment by klonos** [comment]
Not entirely sure what you are trying to achieve, but I think you might get away by creating a custom content type with a separate "h1" filed (text (short)) that you can style how ever you like. Why wouldn't something like that work? ...or is it SEO metatags you are expose to content editors? AFAIK, we haven't yet ported any metatag module to Backdrop.

**Comment by bd0bd** [comment]
**klonos**, thank you!
I need header h1 title, and NOT h1 title which is a default one and there is no option to delete it from the page via admin interface:

![h1](https://cloud.githubusercontent.com/assets/13515302/9883883/6fa192c0-5be5-11e5-9cde-ecf39cfaa0a3.gif)

But also is very actual thing - custom title and h1. Currently there is no way to make it different from each other.

**Comment by bd0bd** [comment]
Here is sample of SEO friendly CMS I have used for many years.

![1](https://cloud.githubusercontent.com/assets/13515302/9884056/61a9ecb6-5be6-11e5-844a-3f7008b12097.gif)

![2](https://cloud.githubusercontent.com/assets/13515302/9884065/6a81962c-5be6-11e5-808f-c94c78332643.gif)

**Comment by bd0bd** [comment]
This is sample of what is needed.
**Articles** has header h1  title:

            <section class="l-section-intro small-12 medium-8 large-8 columns">
            <header>
                <h1>Articles</h1>
            </header>
            </section>

    <article class="alertbox-article-listitem">
        <h1>
            
                <a href="/articles/small-company-intranet/">Should Small Companies Have an Intranet?</a>
            
        </h1>
        <h2>September 13, 2015</h2>
        <p>An intranet can benefit small organizations if they have many remote employees, a high employee turnover, and low findability of company-related content.</p>
    </article>

    <article class="alertbox-article-listitem">
        <h1>
            
                <a href="/articles/gui-slider-controls/">Slider Design: Rules of Thumb</a>
            
        </h1>
        <h2>September 13, 2015</h2>
        <p>Selecting a precise value using a slider is a difficult task requiring good motor skills, even if the slider is well designed. If picking an exact value is important to the goal of the interface, choose an alternate UI element.</p>
    </article>

**Comment by docwilmot** [comment]
Hi there @bd0bd,
You can actually modify the display most things in Backdrop using an override of the theme or a custom layout. 

Also it isnt clear which "title" you are talking about as there are a few. There's the site header title, page title, the node titles.

For your purposes, the I could suggest the following:

Keep in mind that what I am suggesting here is a bit of sacrilege as I am suggesting modifying things directly rather than subtheming (see https://api.backdropcms.org/themes). Instead of just modifying your theme's template files, you should make a subtheme of your theme and modify your subtheme. That way nothing gets accidentally deleted during upgrades. That is the proper way. But this is easier to understand:

- To change the page title, you need to override the layout (template). If you are not using a custom layout, check which template you are using (go to the settings page for the Default layout and cee if it is the two-column, one-column etc). Then just copy that template from `core/layouts` to `/layouts`. For example copy `/core/layouts/one_column` to `/layouts/one_column`. Clear cache and Backdrop will immediately start using that copy of the template instead. Then you can check in that layouts `tpl.php` file and find where the page title line is 
```
        <h1 class="title" id="page-title">
          <?php print $title; ?>
        </h1>
```
and modify it to suit your purposes.

- Similar thing for the article titles, but your theme manages this. Go to your theme (or better, your subtheme), check in the templates folder, and see if there is a `node.tpl.php`. Look for the line that builds node titles and change that:
```
  <header>
    <?php print render($title_prefix); ?>
    <?php if (!$page): ?>
      <h2><a href="<?php print $node_url; ?>"><?php print $title; ?></a></h2>
    <?php endif; ?>
```
Those titles will be changed for all your articles, pages etc.

- For the website title, find the header block template in `core/modules/system/templates` and copy that also to your theme templates folder, and modify that.

**Comment by quicksketch** [comment]
There are a couple of things being requested here. Right now Backdrop doesn't support them, but it could be done with a few module ports:

>   Articles has header h1 title:

To do this, you need to override the node.tpl.php file in your theme. See https://api.backdropcms.org/themes. Then you can have an H1 title on each node, instead of the `<h2>` that each one uses by default.

Some other modules that might be handy in the future: Perhaps a port of https://www.drupal.org/project/page_title. For all other meta tags (description, og:title, keywords, etc.) we need a port of  https://www.drupal.org/project/metatags

**Comment by bd0bd** [comment]
>
To do this, you need to override the node.tpl.php

No.
node.tpl.php is not exactly related to this matter (there is only h2 I can change to h1 within articles tag)
I do not need H1 title on each node.
I need HEADER H1 title on individual pages and H1 title on all other pages.
I talk about page title - **>h1 class="title" id="page-title"<**.

**Comment by docwilmot** [comment]
@bd0bd please see my note above @quicksketch's note (he posted while I was typing).

**Comment by bd0bd** [comment]
>
For the website title, find the header block template in core/modules/system/templates and copy that also to your theme templates folder, and modify that.

I need custom titles for each page and not only for few ones - 1, 2 or 3 pages.

**Comment by docwilmot** [comment]
> I need custom titles for each page and not only for few ones (1, 2 or 3 pages).

Those changes would apply to all pages.

> There is no any tpl.php file inside /layouts/one_column.

Youre right. The `one_column` layout inherits the default template in `core/layouts/templates/layout.tpl.php`. Copy that to your new copy of one_column.

**Comment by bd0bd** [comment]
**docwilmot**, thank you very much for trying to help me!

>
Those changes would apply to all pages.

Probably misunderstanding.
Currently we have similar title for **>h1<** and for **>title<** when feeling the Title * field in.
I need a field within admin interface to make custom **>title<** for every (any) page.
If this field is empty then >title< would be automatically formatted as it is now.
But if we feel this field in then we got custom >title<.

**SAMPLE 1**

http://www.nngroup.com/articles/

**>title<** is

    <title>Usability &amp; UX Articles from Nielsen Norman Group</title>

**>h1<** is:

    <h1>Articles</h1>

**SAMPLE 2**

http://www.nngroup.com/about/

**>title<** is

    <title>Usability &amp; UX Articles from Nielsen Norman Group</title>

**>h1<** is:

    <h1>About NN/g</h1>

**SAMPLE 3**

http://www.nngroup.com/about/

**>title<** is

    <title>Usability and UX Consulting Services | Nielsen Norman Group</title>

**>h1<** is:

    <h1>Consulting</h1>

**Comment by docwilmot** [comment]
Sorry, I truly do not understand what you are saying here.
Maybe draw a mockup of how the full pages would look like, with more explanation.

**Comment by bd0bd** [comment]
I need option to make custom titles for any pages.
Currently it makes similar titles for h1 and for title tag.
This is what I need:
![title](https://cloud.githubusercontent.com/assets/13515302/9904822/922ba944-5c89-11e5-84bd-61182e096966.gif)

**Comment by bd0bd** [comment]
I have gave good sample of CMS (pls look above).
Ok, here is a cropt:

![title](https://cloud.githubusercontent.com/assets/13515302/9904980/9c088832-5c8a-11e5-87c6-3755fe54df81.gif)

**Comment by bd0bd** [comment]
>
Hiding the H1 will invalidate the whole page.

It is a business of a web developer to decide what and when to hide.
We can hide it even without this special option.
Sometimes we need to hide H1 in that place and nothing invalid will happen - please look at this sample: http://www.nngroup.com/ - there is no page title. How I can do this with Backdrop? We have no  normal and easy way to do it. The only way - make it like a block.

**Comment by bd0bd** [comment]
![title](https://cloud.githubusercontent.com/assets/13515302/9905905/4dd7181c-5c90-11e5-8223-c8d0a4039ae2.gif)

![title2](https://cloud.githubusercontent.com/assets/13515302/9905910/535fa01a-5c90-11e5-9c31-bab870858894.gif)

**Comment by quicksketch** [comment]
Thanks for the pictures @bd0bd. From the looks of things, this would be solved for you if we were to port the https://www.drupal.org/project/page_title module to Backdrop, which provides exactly the functionality you're describing. We don't have this module for Backdrop (yet), but it can be done with only a little bit of effort.

**Comment by bd0bd** [comment]
@quicksketch, I think this option should be in the core and not like a separate module.
This is very important thing for sites if anyone wants to make a normal site.
Page title - >title< - is important element for search engines I think you (and other people who is not a novice) know this very well/
Thak you!

**Comment by quicksketch** [comment]
>  canonical tag text field.

We generate a canonical tag already that maps to whatever latest alias is created. We do have a plan to offer redirection on top of this, see #868.

As for the others (meta tags and the head `<title>` element), I'm not entirely sure this should be core functionality. It seems like it's understood infrequently and is so invasive into the UI that it should be a separate add-on module.

In any case, I've updated the issue title to reflect what's being requested in this particular issue, which is the ability to manipulate the head `<title>` independent of the page `<h1>` title.

**Comment by bd0bd** [comment]
>
I'm not entirely sure this should be core functionality

It depends what kind/sort of users (who use Barckdrop) you are targeting.
The CMS, I have used several years, has these options in the core (and they are enabled) - and these are important elements why web-developers choose to buy this CMS. This CMS was marketed within specialized forum for web-developers and SEO-ers.

**Comment by quicksketch** [comment]
> This CMS was marketed within specialized forum for web-developers and SEO-ers.

Yeah our "audience" is still a bit on the vague side: https://backdropcms.org/philosophy

We're trying to fill a gap in between WordPress (blogs) and Drupal 8 (enterprise). Having tons of SEO knobs and whistles may or may not be appropriate for a lot of sites.

Something we can use to guage the appropriateness right now is the usage statistics on the current Page Title module for Drupal 7.

https://www.drupal.org/project/usage/page_title

There are currently 89,028 installs on Drupal 7. And there are 1,009,330 Drupal 7 sites. This means that about 9% of Drupal sites have found this functionality desirable enough that they have installed this module.

So if we follow our philosophy on this particular feature:

> Backdrop core should only include features and tools that benefit the majority of sites that are running it.

We would not include it because it doesn't come anywhere close to being a "majority" feature.

That may not be entirely fair because more users may use the feature if it *were* included in core and they didn't have to install it separately, but it seems pretty clear in this case that although a popular feature, it's not a majority one.

**Comment by bd0bd** [comment]
@quicksketch, I understand you.

> Having tons of SEO ...
No any tons.
I talk only about basic ones which are for granted for any normal full site.
It is not something extravagant or like a new, all these things were actual for a long time over here for websites.

Thank you!

**Comment by klonos** [comment]
At some point there was a discussion about converting the title to a proper field so that it could be hidden because it is currently a "Drupalism" ...and inherently now a "Backdropism" as it is the only field shown on the field management UI ("Manage display" tab) that cannot be hidden. I just cannot remember if that discussion was over at d.org or here and what was the result of it.

I recall that after reading and following that discussion, I was leaning towards converting the title to a proper field. That would solve part of this issue here I believe.

**Comment by klonos** [comment]
...there it is here Backdrop issue: #933

**Comment by jenlampton** [comment]
I think it's too early to know if we need a page_title module in core.  Let's wait until 1) after it's ported and 2) after we have some usage statistics to see if it is worthwhile to 80% of sites running Backdrop.  Closing this issue until then.
